rb_tree.py

```diff
@@ -1,81 +1,287 @@
             return True
 
+        # Case 2: Do we have the same parent? If either we or they are
+        # an orphan/root node, then we only need to check the other's 
+        # parents. Otherwise we need same parent and same value and color.
         if self.parent is None or other.parent is None:
             parents_are_same = self.parent is None and other.parent is None
         else:
-            parents_are_same = self.parent.value == other.parent.value and self.parent.color == other.parent.color
-        return self.value == other.value and self.color == other.color and parents_are_same
+            parents_are_same = ( self.parent.value == other.parent.value and 
+                        self.parent.color == other.parent.color )
+
+        return ( self.value == other.value and 
+                 self.color == other.color and 
+                 parents_are_same )
+
 
     def has_children(self) -> bool:
         """ Returns a boolean indicating if the node has children """
+
         return bool(self.get_children_count())
 
+
     def get_children_count(self) -> int:
         """ Returns the number of NOT NIL children the node has """
-        if self.color == NIL:
+
+        if self.color == NODE_COLOR.NIL:
             return 0
-        return sum([int(self.left.color != NIL), int(self.right.color != NIL)])
+        return sum([int(self.left.color != NODE_COLOR.NIL), int(self.right.color != NODE_COLOR.NIL)])
 
 
-class RedBlackTree:
-    # every node has null nodes as children initially, create one such object for easy management
-    NIL_LEAF = Node(value=None, color=NIL, parent=None)
+NIL_LEAF = Node(value=None, color=NODE_COLOR.NIL, parent=None)
 
-    def __init__(self):
+class RedBlackTree:
+    """
+    This is an implementation of the data structure described in the paper
+    Leonidas J. Guibas and Robert Sedgewick (1978). "A Dichromatic 
+    Framework for Balanced Trees". Proceedings of the 19th Annual 
+    Symposium on Foundations of Computer Science. pp. 8–21. 
+            doi:10.1109/SFCS.1978.3
+    """
+
+    def __init__(self) -> None:
         self.count = 0
         self.root = None
+        self.type = type(None)
+
+        # Used for deletion and uses the sibling's relationship with 
+        # his parent as a guide to the rotation
         self.ROTATIONS = {
-            # Used for deletion and uses the sibling's relationship with his parent as a guide to the rotation
             'L': self._right_rotation,
             'R': self._left_rotation
         }
 
-    def __iter__(self):
-        if not self.root:
-            return list()
+
+    def __iter__(self) -> Node:
+        """
+        Work our way through the Nodes by invoking
+        their __iter__ functions.
+        """ 
+
+        if not self.root: return []
         yield from self.root.__iter__()
 
-    def add(self, value):
+
+    def iter_dump(self) -> str:
+        if not self.root: return [""]
```

Comment:
Let's please follow PEP8 and have the return statement be on the next line. Also empty strings should use single quotes `''`